### PR TITLE
(GH-628) Add server side validation of reCaptcha input

### LINKED ISSUE #628: Contact Maintainers form: captcha not working, maintainers receiving spam e-mails.
I'm receiving spam messages via the Contact Maintainers form.

Below are two examples:

> From: noreply=chocolatey.org@mg.chocolatey.org <noreply=chocolatey.org@mg.chocolatey.org> On Behalf Of 'xrumer888' via Chocolatey.org
> Sent: Tuesday, January 8, 2019 9:58 AM
> To: bcurran3 <RECACTED>
> Subject: [Chocolatey.org] Message for maintainers of the package 'vmware-tools'
> 
> User <xrumer888@outlook.com> sends the following message to the maintainers of Package 'vmware-tools'.
> Gigi Hadid appeared to model Asians inside Instagram the ones Are very angry
> into a thus deleted graphic contributed over the past weekend when his younger mother bella, Gigi Hadid laughs <a href=https://500px.com/asiameofficial>asiame.com</a&gt; and after that squints your loving if supporting a biscuit which appears to achieve face on the samurai no Buddha.
> bella swan Hadid scoured motion picture by way of your girlfriend Instagram credit account, but also forums customer zaynzayns, a follower be aware of Gigi Hadid's boyfriend, singer Zayn <a href=https://asiamecom.tumblr.com/>asiame</a&gt; Malik, Reposted the software. It's unclear <a href=https://500px.com/asiamescam>asiame review</a> exactly how Hadid turned out to be trying to have along with her rapid evolution, But lots of were immediately within consider because racist hostile to cookware people. many individuals found it above all insensitive mainly because Malik is a component oriental: the former One accuracy star's pa is britain Pakistani.
> foods people labeling asains a bit more when individuals allow exciting of people. and also they suggested the chance that Hadid may possibly not have thought of the dessert appeared to be an asian distinct person. Gigi Hadid's distributors to help give this Huffington Post's ask for ideas.
> some people, bella Hadid's telecasting along with its pursuing backlash appeared in the same holiday weekend vogue china and tiawan let go its March manage and features Gigi Hadid.
> Package Url: https://chocolatey.org/packages/vmware-tools
> Pro-tip: It is recommended to (also) post your answer to the vmware-tools comment section so that the information is publicly available.
> This can prevent you from receiving the same question from multiple users.
> Package comments: http://chocolatey.org/packages/vmware-tools#disqus
> ________________________________________
> 

and

> From: noreply=chocolatey.org@mg.chocolatey.org <noreply=chocolatey.org@mg.chocolatey.org> On Behalf Of 'xrumer888' via Chocolatey.org
> Sent: Monday, January 7, 2019 8:05 PM
> To: bcurran3 <REDACTED>
> Subject: [Chocolatey.org] Message for maintainers of the package 'vmware-tools'
>  
> User <xrumer888@outlook.com> sends the following message to the maintainers of Package 'vmware-tools'.
> any data mean Republicans produce a kind downside
> "To an amazing extent, each Republican basic comes with features that happen to be animated just racial discrimination.
> Racial disposition is incredibly challenging measure. you can not simply just inquire everyone if they are hurtful because most people or not satisfying you must presume no. which is partially towards the bias are inclined to performers for your subconscious quantity (want to have proof this? start go ahead and take implied tendency test manufactured by study workers by Harvard, some or even regarding california, since university or college related to virginia).
> and so pollsters frequently attempt and check prejudice not directly. a related force ballot managed in 2012 attempted to calculate implicit racismamong dems in addition Republicans by just inquiring answerers to compare schokohrrutige, white color, hard anodized cookware and after that Hispanic facets. It considered that most 55 for each concerning dems said implied anti- brown thought patterns, dissimilar to 64 per cent together with Republicans an improvement of the fact that lead individuals classified "heavily massive,
> another proxy when considering calibrating racial bias is peoples thoughts about interracial seeing or life. generally Pew investigating town ended up being polling about this focus to find 25 years old. here in 2012, 18 proportion of the Republicans disapproved blues not to mention white wines online dating each other, in comparison to 5 proportion to do with democrats.
> it mat be <a href=https://asiamecom.wordpress.com/>asiame.com</a&gt; more revealing is that this gap consists of widened realize it's a huge. on the inside 1997, essentially the exact same amounts of democrats and as a consequence Republicans <a href=https://www.facebook.com/asiameofficial/>asiame</a&gt; did not approve of interracial relationships, when 30 as well as,while 31 p. c, Respectively. Democratic disapproval of interracial a new relationship offers you melted faster together with Republican disapproval subsequently.
> extremely, stolen might package advantage, Israel's assertion has some carry: data files claim that a area of travelers who actually determine as Republicans specific quantifiable racial prejudice at a higher rate when compared to what an individual's Democratic counterparts.
> This does not mean that all Republicans may be hurtful, of course, or a that most democrats are free of racism. certainly, as the Harvard implicit tendency test demonstrates to, we are all bringing covering racial suitcase using some sort or other or an additional. But laptop computer results around point which will Republicans have an overabundance of work opportunities to do to address these sorts of prejudice <a href=http://asiamescam.weebly.com/>asiame scam</a> within their rating. there's always an obvious moral key the rear until this, But there's also a organizing one: a celebration couldn't persevere on it's own when it developing older and whiter while other nation cultivates younger and varied.
> Package Url: https://chocolatey.org/packages/vmware-tools
> Pro-tip: It is recommended to (also) post your answer to the vmware-tools comment section so that the information is publicly available.
> This can prevent you from receiving the same question from multiple users.
> Package comments: http://chocolatey.org/packages/vmware-tools#disqus

### PR DESCRIPTION
**Pull request by gep13**
Fixes #628

### CODE REVIEW

File: chocolatey/Website/Controllers/PackagesController.cs [hunk 1]
```diff
@@ -554,6 +560,12 @@ public virtual ActionResult ContactAdmins(string id, string version)
         public virtual ActionResult ContactAdmins(string id, string version, ReportAbuseViewModel reportForm)
         {
             if (!ModelState.IsValid) return ContactAdmins(id, version);
+            var response = Captcha.ValidateCaptcha(Request["g-recaptcha-response"]);
+            if (!response.Success)
+            {
+                TempData["ErrorMessage"] = string.Format("Unable to correctly valid reCaptcha: {0}", response.ErrorMessage[0]);
+                return RedirectToAction(MVC.Packages.ContactAdmins(id, null));
```
Comment (ferventcoder) [inline chocolatey/Website/Controllers/PackagesController.cs]: Same notes as above

File: chocolatey/Website/Controllers/PackagesController.cs [hunk 2]
```diff
@@ -600,6 +612,12 @@ public virtual ActionResult ContactOwners(string id)
         public virtual ActionResult ContactOwners(string id, ContactOwnersViewModel contactForm)
         {
             if (!ModelState.IsValid) return ContactOwners(id);
+            var response = Captcha.ValidateCaptcha(Request["g-recaptcha-response"]);
+            if (!response.Success)
+            {
+                TempData["ErrorMessage"] = string.Format("Unable to correctly valid reCaptcha: {0}", response.ErrorMessage[0]);
+                return RedirectToAction(MVC.Packages.ContactOwners(id, null));
```
Comment (ferventcoder) [inline chocolatey/Website/Controllers/PackagesController.cs]: And here

File: chocolatey/Website/Controllers/PagesController.cs [hunk 3]
```diff
@@ -119,7 +120,13 @@ public ActionResult ContactUs()
         public virtual ActionResult ContactUs(ContactUsViewModel contactForm)
         {
             if (!ModelState.IsValid) return View("~/Views/Pages/ContactUs.cshtml", contactForm);
-            
+            var response = Captcha.ValidateCaptcha(Request["g-recaptcha-response"]);
+            if (!response.Success)
+            {
+                TempData["ErrorMessage"] = string.Format("Unable to correctly valid reCaptcha: {0}", response.ErrorMessage[0]);
```
Comment (ferventcoder) [inline chocolatey/Website/Controllers/PagesController.cs]: and here

File: chocolatey/Website/Controllers/PagesController.cs [hunk 4]
```diff
@@ -169,6 +176,12 @@ public ActionResult Discount()
         public virtual ActionResult Discount(DiscountViewModel discountForm)
         {
             if (!ModelState.IsValid) return View("~/Views/Pages/Discount.cshtml", discountForm);
+            var response = Captcha.ValidateCaptcha(Request["g-recaptcha-response"]);
+            if (!response.Success)
+            {
+                TempData["ErrorMessage"] = string.Format("Unable to correctly valid reCaptcha: {0}", response.ErrorMessage[0]);
```
Comment (ferventcoder) [inline chocolatey/Website/Controllers/PagesController.cs]: and here

File: chocolatey/Website/Controllers/UsersController.cs [hunk 5]
```diff
@@ -139,6 +140,12 @@ public virtual ActionResult Register(RegisterRequest request)
             // TODO: consider client-side validation for unique username
 
             if (!ModelState.IsValid) return View("~/Views/Users/Register.cshtml");
+            var response = Captcha.ValidateCaptcha(Request["g-recaptcha-response"]);
+            if (!response.Success)
+            {
+                TempData["ErrorMessage"] = string.Format("Unable to correctly valid reCaptcha: {0}", response.ErrorMessage[0]);
```
Comment (ferventcoder) [inline chocolatey/Website/Controllers/UsersController.cs]: And here

File: chocolatey/Website/Infrastructure/Captcha.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        }
+
+        [JsonProperty("error-codes")]
+        public List<string> ErrorMessage
+        {
+            get;
+            set;
+        }
+    }
+}
```
Comment (ferventcoder) [inline chocolatey/Website/Infrastructure/Captcha.cs]: You might need a new line at the end of the file?

File: chocolatey/Website/Infrastructure/Captcha.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+using System.Web.Configuration;
+using Newtonsoft.Json;
+
+namespace NuGetGallery.Infrastructure
+{
+    public static class Captcha
+    {
+        public static CaptchaResponse ValidateCaptcha(string response)
+        {
+            using (var client = new WebClient())
```
[the remark this answers is not in the document]
Comment (gep13) [reply chocolatey/Website/Infrastructure/Captcha.cs]: Nope, no RestSharp.

**Comment by gep13** [comment]
@ferventcoder can you please review this again?

File: chocolatey/Website/Controllers/PackagesController.cs [hunk 8]
```diff
@@ -508,6 +508,12 @@ public virtual ActionResult ReportAbuse(string id, string version)
         public virtual ActionResult ReportAbuse(string id, string version, ReportAbuseViewModel reportForm)
         {
             if (!ModelState.IsValid) return ReportAbuse(id, version);
+            var response = Captcha.ValidateCaptcha(Request["g-recaptcha-response"]);
+            if (!response.Success)
+            {
+                TempData["ErrorMessage"] = string.Format("Unable to correctly valid reCaptcha: {0}", response.ErrorMessage[0]);
```
[the remark this answers is not in the document]
Comment (ferventcoder) [reply chocolatey/Website/Controllers/PackagesController.cs]: That's interesting.

File: chocolatey/Website/Controllers/PagesController.cs [hunk 9]
```diff
@@ -165,7 +166,7 @@ public ActionResult Discount()
 
         readonly Regex _studentEmailAddressRegex = new Regex(@".*\.edu(\.\w{2})?$|.*\.ac.uk$|.*k12\.\w{2}\.us$", RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);
 
-        [HttpPost, ValidateAntiForgeryToken]
+        [HttpPost, ValidateAntiForgeryToken, Recaptha]
```
Comment (ferventcoder) [inline chocolatey/Website/Controllers/PagesController.cs]: Spelling - it's "Recaptcha" - I wouldn't use that though as the attribute name. I would prefer something like "ValidateFormResponse"

File: chocolatey/Website/Infrastructure/IRecaptchaModel.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace NuGetGallery.Infrastructure
+{
+    interface IRecaptchaModel
```
Comment (ferventcoder) [inline chocolatey/Website/Infrastructure/IRecaptchaModel.cs]: I would call this something more generic - like ISpamValidationModel - this keeps the technology being used out of the name.

File: chocolatey/Website/Infrastructure/IRecaptchaModel.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        string Recaptcha { get; set; }
```
Comment (ferventcoder) [inline chocolatey/Website/Infrastructure/IRecaptchaModel.cs]: Same here on naming.

File: chocolatey/Website/Infrastructure/RecapthaAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+using System;
+using System.Web.Mvc;
+
+namespace NuGetGallery.Infrastructure
+{
+    /// <summary>
+    /// Validates recaptcha responses on server side; apply to ASP.NET MVC action methods.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+    public class RecapthaAttribute : FilterAttribute, IAuthorizationFilter
```
Comment (ferventcoder) [inline chocolatey/Website/Infrastructure/RecapthaAttribute.cs]: You caught my note earlier on spelling and naming. "Recaptcha" not "Recaptha", and I don't like technology names to bleed too much into the model. Recaptcha is something we are using to verify that we are not being spammed. The attribute name should reflect the function, not the technology name. 

Yes, we are using Recaptcha, but if that ever changes, the only code that should change is what is in here versus needing to rename this function and all of the places it is being used.

File: chocolatey/Website/Controllers/PagesController.cs (hunk 9 above)
Comment (ferventcoder) [reply chocolatey/Website/Controllers/PagesController.cs]: Or ValidateAntiSpamToken

File: ChocolateyGallery.sln.DotSettings [hunk 13]
```diff
@@ -86,7 +89,10 @@ limitations under the License.</s:String>
...
 	<s:Boolean x:Key="/Default/Environment/GenerateMru/SortByName/=Implementations/@EntryIndexedValue">False</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/GenerateMru/SortByName/=Overrides/@EntryIndexedValue">False</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpKeepExistingMigration/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpPlaceEmbeddedOnSameLineMigration/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EAddAccessorOwnerDeclarationBracesMigration/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002ECSharpPlaceAttributeOnSameLineMigration/@EntryIndexedValue">True</s:Boolean>
```
Comment (ferventcoder) [inline ChocolateyGallery.sln.DotSettings]: Looks like these snuck into this commit.

**Pull request merged**